src/js/tls.js

```diff
@@ -48,51 +48,40 @@ function TLSSocket(socket, options) {
 
   native.TlsInit(this, options, secureContext);
   this._socketState = socket._socketState;
+
+  var self = this;
+  process.nextTick(function() {
+    if (socket._writableState.ready && !options.isServer) {
+      self._native_connect(options.servername || options.host || 'localhost');
+      self._native_read(null);
+    }
+  });
 }
-util.inherits(TLSSocket, EventEmitter);
+util.inherits(TLSSocket, Duplex);
 
-TLSSocket.prototype._read = native.read;
-TLSSocket.prototype._write = native.write;
-TLSSocket.prototype._connect = native.connect;
+TLSSocket.prototype._native_read = native.read;
+TLSSocket.prototype._native_write = native.write;
+TLSSocket.prototype._native_connect = native.connect;
 
 TLSSocket.prototype.connect = function(options, callback) {
-  this._connect(options.servername || options.host || 'localhost');
-
   if (util.isFunction(callback)) {
     this.on('secureConnect', callback);
   }
-
-  this._socket.connect(options);
-};
-
-TLSSocket.prototype.write = function(data, callback) {
-  if (!Buffer.isBuffer(data)) {
-    data = new Buffer(data);
+  if (this._socket instanceof net.Socket) {
```

Comment:
Do we need to check this? Connect should not be called unless `_socket` has a connect (tls.connect calls this).